File: PythonCode/ChapterWF/Exercise_1_6_4_b.py
""" 
This script estimates expectation values for r
and p for the ground state of Hydrogen.
It does so using atomic units.

Inputs:
   L       - The extension of the radial grid 
   N       - The number of grid points

The input function is used to provide the input parameters from screen.
"""

# Import the numpy library
import numpy as np

# Numerical grid parameters
L = float(input('Please provide the extension of your grid (in a.u.): '))
N = int(input('Please provide the number of grid points: '))


# Normalized wave functions 
def  PsiFunk(r):
  return 2*r*np.exp(-r)


# Set up grid
r = np.linspace(0, L, N)
Psi = PsiFunk(r)              # Vector with function values
h = L/(N-1)                   # Increment 

# Calculate expectation value for r
MeanX = np.trapz(r*np.abs(Psi)**2, r)               # Mean position
print(f'Mean position: {MeanX:.4e} a.u.')

# Set up vector with Psi'(r)
PsiDeriv = np.zeros(N, dtype=complex)               # Allocate and declare
# End points (assume Psi = 0 outside the interval)
PsiDeriv[0] =  Psi[1]/(2*h)
PsiDeriv[N-1] = -Psi[N-1]/(2*h)
# Estimate the derivative with the midpoint rule
for n in range(1,N-1):
  PsiDeriv[n] = (Psi[n+1]-Psi[n-1])/(2*h)


# Calculate expectation value for p
MeanP = -1j*np.trapz(np.conj(Psi)*PsiDeriv, r)   
print(f'Mean momentum: {np.real(MeanP):.4e} a.u.')
print(f'Imaginary part: {np.imag(MeanP):.4e} a.u.')

File: MatlabCode/ChapterWF/Exercise_1_6_4_b.m
% This script estimates expectation values for r
% and p for the ground state of Hydrogen.
% It does so using atomic units.
%
% Inputs:
%   L       - The extension of the radial grid 
%   N       - The number of grid points

% Numerical grid parameters
L = input('Please provide the extension of your grid (in a.u.): ');
N = input('Please provide the number of grid points: ');

% Provide wave function
PsiFunk = @(r) 2*r.*exp(-r);

% Set up grid
r = linspace(0,L,N);
Psi = PsiFunk(r);               % Vector with function values
h = L/N;                        % Increment 

% Calculate expectation value for r
MeanR = sum(r.*abs(Psi).^2)*h;                 % Mean position
% Write result to screen
disp(['Mean distace: ', num2str(MeanR), ' a.u.'])

% Set up vector with Psi'(r)
PsiDeriv = zeros(1, N);              % Allocate
% End points
PsiDeriv(1) = (Psi(2)-Psi(1))/h;    % Two point formula
PsiDeriv(N) = -Psi(N-1)/(2*h);
% Estimate the derivative with the midpoint rule
for n = 2:(N-1)
  PsiDeriv(n) = (Psi(n+1)-Psi(n-1))/(2*h);
end

% Calculate expectation value for p
MeanP = -1i*sum(conj(Psi).*PsiDeriv)*h;   % Mean momentum
% Write result to screen
disp(['Mean momentum (real part): ', num2str(real(MeanP)), ' a.u.'])
